src/cdomains/intDomain.ml

```diff
@@ -911,30 +944,631 @@ struct
-        let rec max m1 ms = match ms with | [] -> m1 | x::xs -> match m1 with
-              | None -> max (Some x) xs | Some m -> if Ints_t.compare m x > 0 then max (Some m) xs else max (Some x) xs in
-        match min None ls, max None ls with
-        | Some m1, Some m2 -> refine_with_interval ik (Some(l, u)) (Some (m1, m2))
-        | _, _-> intv
+      let rec min m1 ms = match ms with | [] -> m1 | x::xs -> match m1 with
+        | None -> min (Some x) xs | Some m -> if Ints_t.compare m x < 0 then min (Some m) xs else min (Some x) xs in
+      let rec max m1 ms = match ms with | [] -> m1 | x::xs -> match m1 with
+        | None -> max (Some x) xs | Some m -> if Ints_t.compare m x > 0 then max (Some m) xs else max (Some x) xs in
+      match min None ls, max None ls with
+      | Some m1, Some m2 -> refine_with_interval ik (Some(l, u)) (Some (m1, m2))
+      | _, _-> intv
+
+  let project ik p t = t
+end
+
+module IntervalSetFunctor(Ints_t : IntOps.IntOps): SOverflow with type int_t = Ints_t.t and type t = (Ints_t.t * Ints_t.t) list =
+struct
+
+  module Interval = IntervalFunctor(Ints_t)
+
+  let name () = "interval_sets"
+
+  type int_t = Ints_t.t
+
+  let (>.) a b = Ints_t.compare a b > 0
+  let (=.) a b = Ints_t.compare a b = 0
+  let (<.) a b = Ints_t.compare a b < 0
+  let (>=.) a b = Ints_t.compare a b >= 0
+  let (<=.) a b = Ints_t.compare a b <= 0
+  let (+.) a b = Ints_t.add a b
+  let (-.) a b = Ints_t.sub a b
+  let ( *.) a b = Ints_t.mul a b
+  let (/.) a b = Ints_t.div a b
+
+  let min4 a b c d = Ints_t.min (Ints_t.min a b) (Ints_t.min c d)
+  let max4 a b c d = Ints_t.max (Ints_t.max a b) (Ints_t.max c d)
+
+  (*
+    Each domain's element is guaranteed to be in canonical form. That is, each interval contained
+    inside the set does not overlap with each other and they are not adjacent.
+  *)
+  type t = (Ints_t.t * Ints_t.t) list [@@deriving eq, hash, ord]
+
+  let range ik = BatTuple.Tuple2.mapn Ints_t.of_bigint (Size.range ik)
+
+  let top () = failwith @@ "top () not implemented for " ^ (name ())
+
+  let top_of ik = [range ik]
+
+  let bot () = []
+
+  let bot_of ik = bot ()
+
+  let show (x: t) =
+    let show_interval i = Printf.sprintf "[%s, %s]" (Ints_t.to_string (fst i)) (Ints_t.to_string (snd i)) in
+    List.fold_left (fun acc i -> (show_interval i) :: acc) [] x |> List.rev |> String.concat ", " |> Printf.sprintf "[%s]"
+
+  (* New type definition for the sweeping line algorithm used for implementing join/meet functions. *)
+  type event = Enter of Ints_t.t | Exit of Ints_t.t
+
+  let unbox_event = function Enter x -> x | Exit x -> x
+
+  let cmp_events x y =
+    let res = Ints_t.compare (unbox_event x) (unbox_event y) in
+    if res <> 0 then res
+    else
+      begin
+        match (x, y) with
+        | (Enter _, Exit _) -> -1
+        | (Exit _, Enter _) -> 1
+        | (_, _) -> 0
+      end
```

Comment:
I was about to suggest `[@@deriving ord]`, but I'm guessing it's intentionally comparing the ints before the event constructors? If so, then this deserves a comment, so nobody accidentally cleans this up in the future.

Reply:
I'll make sure to add an appropriate comment!

Reply:
Added.